packages/redaction/src/generic/redact-recursively.ts

```diff
@@ -0,0 +1,105 @@
+/**
+ * Given an object recurses through it's structure, including inside arrays, and applies
+ * the provided redaction functions. For example:
+ *
+ * ```js
+ * // Replace all (non-id) strings with asterixes
+ * redactRecursively({
+ *  obj: myObject,
+ *  redactString: (str, path) => path.at(-1) === "id" ? str : str.replace(/./g, "*"),
+ * })
+ * ```
+ */
+export const redactRecursively = (
+  rootValue: object,
+  {
+    redactObject,
+    redactArray,
+    redactString,
+    redactNumber,
+    redactBigInt,
+    redactBoolean,
+  }: {
+    /**
+     * If the original object is returned then redactRecursively will recurse inside it,
+     * and continue applying the redaction.
+     *
+     * If a new object is returned (different memory reference) then redactRecursively will
+     * replace the existing object with the new object and not recurse inside it.
+     */
+    redactObject?: (obj: object, jsonPath: string[]) => object;
+
+    /**
+     * If the original array is returned then redactRecursively will recurse inside it,
+     * and continue applying the redaction.
+     *
+     * If a new array is returned (different memory reference) then redactRecursively will
+     * replace the existing array with the new array and not recurse inside it.
+     */
```

Comment:
```suggestion
    /**
     * If the original array is returned then redactRecursively will recurse inside it,
     * and continue applying the redaction.
     *
     * If a new array is returned (different memory reference) then redactRecursively will
     * replace the existing array with the new array and not recurse inside it.
     *
     * Similar to {@link redactObject} you should make sure that your implement executes
     * quickly in cases where you return the original object unchanged, since it may be executed
     * many times over in such a case. See the failure case example in the JSDoc for {@link redactObject}
     * for more details.
     */
```